### DISCUSSION TITLE
How to use dataloader in mutations

### DISCUSSION THREAD

**Discussion by Bertg** [rmosolgo/graphql-ruby#4492] [Q&A]
Say I have a defined mutation resolver like this:

```
class SoftDeleteMembers < GraphQL::Schema::RelayClassicMutation

  argument :member_ids, [ID], required: true, loads: Types::MemberType, as: members
  field :soft_deleted_members, [Types::MemberType], null: false

  def resolve(members: )
    # Do some deletion logic, that mutates the given members in place

    {
      soft_deleted_members: dataloader.with(Sources::RepoById, MemberRepo).load_all(members.map(&:id))
    }
  end
```

This does not work as expected. The loader used by the argument also uses the same dataloader.

In a lot of cases this is not a problem, rarely do we return the same set of object we are loading, but here it matters a lot.
Can we, somehow instruct data loader to clear its cache? Or should we prevent loading data from the arguments?

@rmosolgo Your guidance on "best practice" here would be invaluable.

**Comment by rmosolgo** [comment]
Hey, sorry I missed this before. 

> This does not work as expected. 

Could you please describe how it _currently_ works, and then how you'd _expect_ it to work? I haven't tried it this way myself and I don't know exactly what you have in mind. 

Maybe related to #4600

**Comment by Bertg** [comment]
@rmosolgo Yeah, I think this is the same as #4600

In the example I posted we see the following:

1.  DataLoader loading the objects when resolving the argument
2. An update to the database may update one or more of the objects loaded in step one in the database
3. When returning the data the dataloader already contains data and won't be reloaded. So it can't pick up the changes from step 2. Thus the data is stale.

I'll close this conversation, and follow up in issue #4600
